### ISSUE TITLE
UseCase_a05_RotatingChildNode kind of slow

### ISSUE THREAD

**Issue by elect86** [nifty-gui/nifty-gui#402]
I am running the following sample `UseCase_a05_RotatingChildNode` from nifty 2.0 and it is not completely fluid

```
      frame     update     render    r.state  r.content     r.pass    # batch      input      total        fps
       7699     0.0000    20.2855     0.0634     4.7725    15.4436          0     0.0091    20.3006    49.2596
       7700     0.0000    20.3791     0.0725     6.9399    13.3637          0     0.0121    20.3972    49.0263
       7701     0.0000    18.6645     0.0694     4.7967    13.7924          0     0.0091    18.6796    53.5344
       7702     0.0000    17.9732     0.1721     4.3771    13.4180          0     0.0091    17.9913    55.5823
       7703     0.0000    20.7263     0.1781     4.5582    15.9869          0     0.0091    20.7444    48.2058
       7704     0.0000    19.2199     0.0755     4.6186    14.5198          0     0.0091    19.3346    51.7206
       7705     0.0030    18.3596     0.1660     4.3922    13.8014          0     0.0091    18.3777    54.4137
       7706     0.0000    17.8797     0.1781     5.0352    12.6664          0     0.0091    17.8947    55.8823
       7707     0.0000    19.1837     0.1660     4.7242    14.2874          0     0.0091    19.1988    52.0866
       7708     0.0000    19.4554     0.1781     5.6208    13.6535          0     0.0091    19.4735    51.3518
```

What could it be? Maybe vsync? How can I check/disable it?

**Comment by bgroenks96** [comment]
Will look into it more ASAP. Thanks for reporting this.

In the mean time, do you have VisualVM installed? If you can run the example and post a screenshot of the CPU profiler, that would be fantastic.

**Comment by void256** [comment]
I'm currently working on it. Going to commit the first changes soon.

There were a couple of issues in the 2.0 renderer. Mostly because we wanted something that just works to continue development in other areas. So it was not optimized at all.

Here are a few issues that are already resolved / improved (about to push this soon):
- GC usage was terrible due to creating a lot of garbage each frame especially in Mat4 matrix/transformation - this is resolved and GC pressure is already reduced a lot
- everything was rendered each frame again without any kind of caching - there is some automatic saving and reusing in textures supported now so when gui elements don't change nothing is rendered
- texture batches allocated memory for 1000 primitives (quads) each and then only stored one textured quad in it :smile: - this hurted performance a lot since there is currently not yet a texture atlas so we had lots of batch switching and allocating `1000*8*6` floats for each batch was not a good idea :grin: I reduced it to one primitive for now (see below)
- I've dropped support to automatically cache child nodes in parent textures - this was more a way to simplifiy the renderer and make it easier to understand and not so much about performance though

This will already get us back to 200 to 400 fps or so for this specific example.

The next step to improve it further is to bring back the texture atlas from 1.x. Nifty. 2.0 is already prepared for it but it's not yet active or finished. So currently we have a lot of texture switching going on which in turn will make our texture batches small (1 quad ;) so this is going to improve next. When we don't need to switch the active texture anymore we can make the texture batches bigger again.

To answer the original question .. this is normal (for now) and the solution would be to be a little more patient and keep an eye on the list of commits :wink:

**Comment by elect86** [comment]
Hi guys, you are great, I am gonna reply deeply tomorrow when I'll be back on it ;)

**Comment by bgroenks96** [comment]
@void256 Whew. I'm glad you knew what was going on. 40-50 fps on a GTX 770M GPU had me freaking out lol.

**Comment by void256** [comment]
btw help is always appreciated :smile: a nice way to start for someone willing to help would be to take care of issue #391 and send a pull request for it :D

**Comment by bgroenks96** [comment]
Or I could fix it and NOT send a pull request :stuck_out_tongue_winking_eye:

**Comment by void256** [comment]
not you! :wink: you're already busy with lots of other things :yum:

**Comment by bgroenks96** [comment]
True. Still haven't gotten around to submitting CoreGL to Sonatype :disappointed:

**Comment by elect86** [comment]
Ehm, it wasn't the mobile version... 770 desktop ^^

Anyway, I (think I) am good at engine stuff, I could take a look at your core rendering maybe

You mentioned the mat4 transformation, which math lib are you using?

Regarding the texture switching, which opengl profile do you target? Because I guess you will benefit enormously from gl4 features (especially ssao and bindless and sparse textures)..

**Comment by bgroenks96** [comment]
> Ehm, it wasn't the mobile version... 770 desktop ^^

That's even worse then!

> You mentioned the mat4 transformation, which math lib are you using?

I think Nifty has its own code for that (either that or CoreGL does).

> Regarding the texture switching, which opengl profile do you target? Because I guess you will benefit enormously from gl4 features (especially ssao and bindless and sparse textures)..

CoreGL targets 3.3+ and Nifty uses CoreGL, so 3.3. Since it's a GUI library, we kind of have to target the widest compatibility base possible. We'll probably have to reduce the GL feature set more actually to support ES3.

**Comment by elect86** [comment]
I am currently rewriting my math lib for high performance rendering (you can avoid literally _any_ allocation if you want)

You can target 3.3+ but if you have available 4.5 why don't exploit it too?

**Comment by bgroenks96** [comment]
We can certainly look into it for the future, but the Nifty 2.0 API is still under heavy development. So it's better to keep the renderer as simple as possible for the time being.

**Comment by elect86** [comment]
~1.7k fps now, much better  :smile: 

thanks

Anyway, a GUI such [this](http://imgur.com/qs2cbYJ) won't be that complicate to realize, will it?

**Comment by relu91** [comment]
Just to have an other reference in my notebook about 320 fps with drop to 22 and once to 4. I guess for a mobile it's super cool :+1: (NVIDIA GeForce GT 540M). I have also notice a drastically drop of GC calls and also less memory usage! Great work @void256 .

> Anyway, a GUI such this won't be that complicate to realize, will it? 

With 2.0 I think you could encounter some difficulties. It's still under development and it doesn't have support for controls ( such as buttons, scroolbars ... etcc ). Right now you can mix nodes have something like that tough.

**Comment by bgroenks96** [comment]
I take it this issue was fixed by 4f09e3b73af9.

In that case, I'll go ahead and close. Feel free to re-open if there are any more issues.

P.S. @elect86 I'm not sure, but you can always ask if you encounter any difficulties. Good luck!
